### PR TITLE
Speed up generation of output

### PR DESCRIPTION
**Pull request by AlCalzone**
This is a proof of concept for a different rendering approach that does not rely on `sliceAnsi` as heavily.

I tried this with my test component from #560, but with 50 spinners. Without this change, 50 spinners put one core of my CPU at 100%, and rendering lags heavily. With this change, I'm looking at ~40% utilization of one core, which is still not ideal, but renders fluently.

The output for 10s of profiling now looks like this, is idle half the time and most of the work is done in `wasm`, which is probably from `yoga` internals:
grafik

fixes: https://github.com/vadimdemedes/ink/issues/560

This will need some more testing with emojis and clipping, but I wanted to get this out for some early feedback.

### CODE REVIEW

**Comment by AlCalzone** [comment]
Ok, looks like emojis and some full-width characters are causing test failures, I'll look into that.

And there's probably an ANSI code in a different position than before - I don't see a visual difference here:
grafik

**Comment by vadimdemedes** [comment]
This is really impressive! Would you be willing to push this PR to completion?

**Comment by AlCalzone** [comment]
Yeah, will do soon.

**Comment by AlCalzone** [comment]
Tests are passing too now. Just for fun, I ran `time npm run benchmark benchmark/simple` to compare.
**master**
```
real    0m11,412s
user    0m13,342s
sys     0m0,604s
```

**this PR**
```
real    0m8,208s
user    0m9,051s
sys     0m0,330s
```

So roughly 32% faster rendering for even simple outputs.

File: src/output.ts [hunk 1]
```diff
@@ -92,10 +98,20 @@ export default class Output {
 
 	get(): {output: string; height: number} {
 		// Initialize output array with a specific set of rows, so that margin/padding at the bottom is preserved
-		const output: string[] = [];
+		const output: StyledChar[][] = [];
```
Comment (vadimdemedes) [inline src/output.ts:101]:
```suggestion
		const output: Array<StyledChar[]> = [];
```

File: src/output.ts [hunk 2]
```diff
@@ -172,29 +188,41 @@ export default class Output {
...
-						sliceAnsi(currentLine, x + width);
+					const chars = styledCharsFromTokens(tokenize(line));
+					let offsetX = x;
+					for (const char of chars) {
+						currentLine[offsetX] = char;
+
+						// Some characters take up more than one column. In that case, the following
+						// pixels need to be cleared to avoid printing extra characters
+						const charWidth = char.fullWidth ? 2 : char.value.length;
+						for (let dx = 1; dx < charWidth; dx++) {
```
Comment (vadimdemedes) [inline src/output.ts:206]: Could you use a descriptive variable name for index instead of `dx`? Does `characterIndex` fit here?

File: src/output.ts [hunk 3]
```diff
@@ -172,29 +188,41 @@ export default class Output {
...
 
 					for (const transformer of transformers) {
 						line = transformer(line);
 					}
 
-					output[y + offsetY] =
-						sliceAnsi(currentLine, 0, x) +
-						line +
-						sliceAnsi(currentLine, x + width);
+					const chars = styledCharsFromTokens(tokenize(line));
```
Comment (vadimdemedes) [inline src/output.ts:198]:
```suggestion
					const characters = styledCharsFromTokens(tokenize(line));
```

File: src/output.ts [hunk 4]
```diff
@@ -172,29 +188,41 @@ export default class Output {
...
+					let offsetX = x;
+					for (const char of chars) {
```
Comment (vadimdemedes) [inline src/output.ts:200]:
```suggestion
					for (const character of characters) {
```

File: src/output.ts (hunk 2 above)
Comment (vadimdemedes) [inline src/output.ts:206]: Why does this skip the first character and starts from the second one?

**Comment by vadimdemedes** [comment]
Superb work 🔥

File: src/output.ts (hunk 2 above)
Comment (AlCalzone) [reply src/output.ts:206]: Because the first char is the char that's actually written, so width 1 is already accounted for.

Comment (vadimdemedes) [reply src/output.ts:206]: Not sure I understand, could you elaborate the "that's actually written" part?

Comment (AlCalzone) [reply src/output.ts:206]: Line 201 writes the current character to the output array. Some characters are wider than a single column though, so we have to set the following columns (up to the character width) to "" so we don't end up with extra characters.
So if the char is a 2-column emoji, we write the emoji to column x and "" to column x+1. The character after the emoji will then correctly be at column x+2.

File: src/output.ts [hunk 5]
```diff
@@ -172,29 +188,41 @@ export default class Output {
...
+
+						// Some characters take up more than one column. In that case, the following
+						// pixels need to be cleared to avoid printing extra characters
+						const charWidth = char.fullWidth ? 2 : char.value.length;
+						for (let dx = 1; dx < charWidth; dx++) {
+							currentLine[offsetX + dx] = {
+								type: 'char',
+								value: '',
+								fullWidth: false,
+								styles: char.styles
```
Comment (AlCalzone) [inline src/output.ts:211]: maybe this should be
```suggestion
								styles: []
```
though.

File: src/output.ts (hunk 2 above)
Comment (vadimdemedes) [reply src/output.ts:206]: Do we need a loop here then? Could this be simplified to this?

```js
if (char.fullWidth) {
  currentLine[offsetX + 1] = /* empty char */
}

offsetX += character.fullWidth ? 2 : 1
```

Comment (AlCalzone) [reply src/output.ts:206]: Aren't there some emojis which have length > 4?

Comment (AlCalzone) [reply src/output.ts]: I've pushed a commit to eliminate the loop.

File: src/output.ts [hunk 6]
```diff
@@ -198,12 +198,16 @@ export default class Output {
 					const chars = styledCharsFromTokens(tokenize(line));
 					let offsetX = x;
 					for (const char of chars) {
+						if (offsetX >= this.width) break;
```
Comment (vadimdemedes) [inline src/output.ts:201]: How did you run into this issue where an out-of-bounds write would occur? Could you add a test for it to ensure it doesn't regress back?

Comment (AlCalzone) [reply src/output.ts:201]: Done. In my case it was a pretty complicated nested layout, but I managed to reduce it to a `<Text wrap="end">` that's longer than the stdout width.

Comment (vadimdemedes) [reply src/output.ts:201]: There's no `wrap="end"` value, but there's `wrap="truncate-end"`. I updated it in the test you added, then commented out "prevent out-of-bounds writes" commit changes and it didn't crash for me. Looks like we don't need this code then?

Comment (AlCalzone) [reply src/output.ts:201]: Hmm maybe the test case isn't correct then. I ended up with gaps in the output array, which crashed the function that converts Tokens back to strong.

Comment (vadimdemedes) [reply src/output.ts:201]: Do you still see the same issue when using `wrap="truncate-end"` instead of `wrap="end"` (which is an invalid prop)?

Comment (AlCalzone) [reply src/output.ts:201]: Will try tomorrow or so.

Comment (vadimdemedes) [reply src/output.ts:201]: Hey @AlCalzone, do you still have the capacity to finish this PR? Would really love to ship it. No worries if you're busy though.

Comment (vadimdemedes) [reply src/output.ts:201]: No worries, take your time 👍

**Comment by vadimdemedes** [comment]
Thanks @AlCalzone, that was really helpful! Here's what I've found.

I took your `Box` example and rendered it in the terminal.

```jsx
import React from 'react';
import {Box, render} from 'ink';

function Test() {
	return <Box width={12} height={10} borderStyle={'round'}></Box>;
}

process.stdout.columns = 10;
render(<Test />, {debug: true});
```

CleanShot 2023-07-16 at 11 13 06@2x

The right border is rendered in the wrong place, even though there are correct X and Y coordinates for all sides:

- Top border has X = 0, Y = 0
- Left border has X = 0, Y = 1
- Right border has X = 11, Y = 1
- Bottom border has X = 0, Y = 9

Here's where each line is written to output:

https://github.com/vadimdemedes/ink/blob/8a0476024cf997e0dbc02e493ad3972349b02474/src/output.ts#L187-L190

I added this `console.log` to see what exactly gets written and where:

```js
console.log({
	// X coordinate of the `line` to write to output
	x,
	// Existing output before X coordinate
	before: sliceAnsi(currentLine, 0, x),
	beforeWidth: sliceAnsi(currentLine, 0, x).length,
	// New text to write to output
	line,
	lineWidth: line.length,
	// Existing output after X coordinate + width of the `line`
	after: sliceAnsi(currentLine, x + width),
	afterWidth: sliceAnsi(currentLine, x + width).length
});
```

Here's the output for one of the lines for writing the right border:

```
{
  x: 11,
  before: '│         ',
  beforeWidth: 10,
  line: '│',
  lineWidth: 1,
  after: '',
  afterWidth: 0
}
```

Normally, `sliceAnsi(currentLine, 0, x)` should return a string where `length === x`, but here `beforeWidth` equals 10, not 11. The reason is `currentLine` comes from the `output` variable, which is initialized with strings that are as wide as the terminal (in our case terminal width is 10).

https://github.com/vadimdemedes/ink/blob/8a0476024cf997e0dbc02e493ad3972349b02474/src/output.ts#L174

https://github.com/vadimdemedes/ink/blob/8a0476024cf997e0dbc02e493ad3972349b02474/src/output.ts#L97-L99

This creates a problem where `output` contains lines with the length of 10, so the index of the last character in each line is 9. At the same time, we're trying to write a character at index 11, which results in invalid output where character at index 10 is missing. As a result, right border is placed at X = 10, not X = 11.

It doesn't cause a crash in the current version of Ink, because `sliceAnsi` is used to extract string ranges, which doesn't fail if you run `sliceAnsi(str, 0, 11)` where `str.length === 10`. However, your PR no longer uses `sliceAnsi` and is sensitive to `undefined` items.

To sum up, this is an existing issue of Ink with content larger than the terminal and should be fixed separately from this PR. To avoid regressions in this PR though, we need to skip `undefined` array items to keep the rendering behavior the same as it is now.

**Comment by vadimdemedes** [comment]
Pushed a commit to keep existing behavior with out of bounds writes.

**Comment by vadimdemedes** [comment]
CI is green, merging!

Thank you for this incredible performance boost, @AlCalzone!
